src/script/engines/lua.c

```diff
@@ -215,6 +215,66 @@ static const int _mScriptSocketNumErrors = sizeof(_mScriptSocketErrors) / sizeof
 
 #if LUA_VERSION_NUM < 503
 #define lua_pushinteger lua_pushnumber
+
+static int _luaPairs(lua_State* lua) {
+	luaL_checkany(lua, 1);
+	if (luaL_getmetafield(lua, 1, "__pairs") == LUA_TNIL) {
+		lua_getglobal(lua, "next");
+		lua_pushvalue(lua, 1);
+		lua_pushnil(lua);
+	} else {
+		lua_pushvalue(lua, 1);
+		lua_call(lua, 1, 3);
+	}
+	return 3;
+}
+
+static int _luaIpairsIter(lua_State* lua) {
+	lua_Integer i = luaL_checkinteger(lua, 2) + 1;
+	lua_pushinteger(lua, i);
+	lua_pushvalue(lua, -1);
+	lua_gettable(lua, 1);
+	return (lua_type(lua, -1) == LUA_TNIL) ? 1 : 2;
+}
+
+static int _luaIpairs(lua_State* lua) {
+	luaL_checkany(lua, 1);
+	lua_pushcfunction(lua, _luaIpairsIter);
+	lua_pushvalue(lua, 1);
+	lua_pushinteger(lua, 0);
+	return 3;
+}
```

Comment:
Isn't pairs/ipairs only in 5.2/5.3, not 5.1?

Reply:
No, they're in 5.1. You're thinking of the metamethods.

Reply:
That's right. In fact these modifications add __pairs metamethod to Lua 5.1, and removes the LUA_TTABLE type restriction for pairs/ipairs.

Reply:
What's the use case? (And if you're implementing this, why not add __ipairs too?)

Reply:
-Lua 5.1: pairs and ipairs work only with table objects. Also they don't check for __ respective metamethods.
-Lua 5.2: pairs and ipairs work only with table objects. They check for __ respective metamethods.
-Lua 5.3: pairs check for __pairs metamethod. ipairs check for __ipairs metamethod only in compatibility mode. Object type is irrelevant (only checks if its a table when using luaB_next iterator (the global next function).
-Lua 5.4: Only pairs check for its __ metamethod. Like in 5.3 object type is irrelevant (only checks if its a table when using luaB_next iterator (the global next function).

So the approach is done with Lua 5.4's functionality in mind.

It's used for mSTList iteration and mSTTable __pairs metamethod.

Reply:
If you're trying to support 5.4 semantics then you need to add support for __len to the ipairs implementation.

Reply:
Oops... You're right!

I'll fix that.

Reply:
Okay, this seems different... I think there's no easy way of overriding # operator, neither is a replaceable function like pairs and ipairs.

Reply:
You're right, there isn't, but if you're going to provide the 5.4 semantics for ipairs then you can check for the metamethod manually.

Reply:
I think I'm not getting this... What connection is between __len and ipairs? ipairs in Lua 5.4 is different from Lua 5.1, but I think I've adjusted it propperly in _luaIpairsIter and _luaIpairs functions.

Reply:
*checks documentation* Oh. Huh. I was under the impression that ipairs was semantically equivalent to `for i = 1, #t do` in 5.4 since the __ipairs metamethod was removed, but apparently I was mistaken.